### PR TITLE
Update contributing.md with gathered notes from top contributors on their expectations for new committers

### PR DESCRIPTION
**Pull request by FaultyJuggler**
Philip Bell closing the loop on our efforts to update and standardize the documentation on how to commit to Presto about a year ago. Updates are based on our "How to Commit to Presto" group discussions and working group between Ahana, Meta, Uber, and top contributors.

## Description
Using the rich text document produced for presenting clear expectations for committers and migrated the information as best as possible to the contributing.md in markdown.

## Motivation and Context
Previously the community expressed repeated frustration that expectations to commit were not explicit, excited contributors would spend hours on fixes and features only to be left waiting a long time to be told they did not meet some criteria that had not been communicated explicitly or in detail in the existing documentation.

The updates to the contributing.md hopes to resolve some of these issues.

```
== RELEASE NOTES ==

General Changes
* contributing.md now matches expectations defined by top contributors as of mid 2022
```

### CODE REVIEW

File: CONTRIBUTING.md
Comment (yhwang) [inline CONTRIBUTING.md:135]:
```suggestion
* Add or modify existing tests related to code changes being submitted
```

File: CONTRIBUTING.md
Comment (yhwang) [inline CONTRIBUTING.md:225]: 1. private ?

File: CONTRIBUTING.md
Comment (yhwang) [inline CONTRIBUTING.md:255]: reduce one level of indentation. I guess this should be the same as line 246
the thing applies to line 262, 269, and 278

File: CONTRIBUTING.md
Comment (yhwang) [inline CONTRIBUTING.md:309]:
```suggestion
    * Creating Optional objects is not free. If the function call is in a critical loop, do NOT use Optional parameters.
```

File: CONTRIBUTING.md
Comment (yhwang) [inline CONTRIBUTING.md:357]: yea, Java 8

**Comment by FaultyJuggler** [comment]
@yhwang I have committed the changes you requested

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:147]:
```suggestion
        1. No data structures that can grow without bounds. Memory usage is accounted for
```

I will create an issue for adding this to our functions doc--very important to get right, and it's missing!  CC: @ZacBlanco

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:153]: Perhaps mention that large features may need an RFC: https://github.com/prestodb/rfcs

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:237]:
```suggestion
    * When it's not possible, ensure that accesses to the non-final fields are thread safe by whatever methods are appropriate for the circumstance (Concurrent collections, synchronized access, etc.) if they are in a code path that is multi-threaded
```

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:426]:
```suggestion
    * A PR can consist of multiple small commits, preferably not exceeding 20.
```

We can keep it like this for now and adjust if needed

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:427]:
```suggestion
    * The total number of lines modified in a single PR shall not exceeding 5000, except for changes that include checked in code generated files (such as [presto_protocol.cpp](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/presto_protocol/presto_protocol.cpp)).
```

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:438]: I think we can simplify this and simply instruct people to follow the PR template, which has check boxes one must check.

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:450]: I think this can be simplified, since even committers must have a fork and then submit a PR from their fork.

It might make sense to link to the fork and pull model, since that is what we follow: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/about-collaborative-development-models#fork-and-pull-model

**Review by tdcmeehan — COMMENTED** [review]
This is great!  Some minor suggestions.

**Comment by tdcmeehan** [comment]
CC: @steveburnett

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:341]:
```suggestion
    * Inline function calls when appropriate. For example, if a function is only called once, we don’t need to create a variable for it.
```

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:343]:
```suggestion
      For example, Instead of doing this
```

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:375]:
```suggestion
          For example, “If applied, this commit will ___________”
```

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:382]:
```suggestion
        * Commits should pass tests on their own, not be dependent on other commits
```

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:400]:
```suggestion
      For example,\
      Resolves: #1234\
      See also: #1235, #1236
```

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:500]:
```suggestion
* Review code with the objective of helping someone land their changes
```

File: CONTRIBUTING.md
Comment (kiersten-stokes) [inline CONTRIBUTING.md:433]: It might be worth linking to the relevant section here or maybe just add the word "above" somewhere in this line

```suggestion
    * Please refer to the section on commit subject
```

**Review by kiersten-stokes — COMMENTED** [review]
This will be super helpful for new contributions! I just have a few NITs, most of which are the same comment

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:177]: Nice work providing an example of incorrectly broken down code for the reader to compare their code to. No change requested here, this is another callout of good work.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:314]:
```suggestion
    * Recommend to add the “/** */” (Javadoc) style comments to all interface methods, with explanation of the parameters and returns. For example:
```

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:373]:
```suggestion
        * Use the body to explain what and why versus how
```
GitLab doc style guide doesn't mention vs. or versus. The [Google developer doc word list entry for vs.](https://developers.google.com/style/word-list#vs) recommends using the unabbreviated version.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:377]:
```suggestion
        * Use the body for What and Why, 
        * If your commit is complex or dense, share some of the How context in the body
```
This felt like it could be two separate guideline points: let me know if you feel strongly that this should stay one line.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:388]:
```suggestion
            * Do not assume the code is self-evident or self-documenting.
```
Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:397]:
```suggestion
    * If the commit was to solve a Github issue, refer to it at the end of a commit message in a rfc822 header line format.\
```

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:412]:
```suggestion
    * Try to follow the original PR and commits structure. Do not mingle multiple original commits or reorder them.
```
Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:448]:
```suggestion
If you are not an official committer, you must fork from Master and then submit a Pull Request to contribute as an individual.
```

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:478]: Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:482]:
```suggestion
    * Implement or modify relevant tests, otherwise provide clear explanation why test updates were not necessary
```
Suggest "explanation" to avoid using "documentation" in a way different than the (user-facing) documentation being discussed in the immediately previous lines, to reduce ambiguity.

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:493]: I don't understand the meaning of the indent for "Review the related code, draft initial documentation as a separate commit". 
Is this another "what not to do", or a recommendation of what you should do instead of "Protesting lack of documentation for a code section"?

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:494]:
```suggestion
* Submit without test cases or clear justification for lack thereof
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for this! 
I made many comments but like @kiersten-stokes's feedback, almost all of my comments are nits or very minor issues that are generally repetitions of the same couple of comments, so please don't take the quantity of my comments as a negative opinion of the work. I also made comments to emphasize and compliment bits that I thought were excellent.

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md:134]: Will get those added!

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md]: Good catch

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md:140]: Agreed, will make the change

File: CONTRIBUTING.md
Comment (jaystarshot) [reply CONTRIBUTING.md:125]: Not sure this is needed for each issues

Comment (tdcmeehan) [reply CONTRIBUTING.md:125]: Let's attempt to break this down more precisely using T-shirt sizes:

1) Large chances should have an RFC.  The RFC should be reviewed before patches are submitted.
2) Medium size changes should have an issue.  Work from RFCs may be broken down into smaller issues, which themselves link to the RFC.
3) Smaller changes, such as minor bug fixes and code formatting, may forego an issue and just submit a PR.

The sizing of each will require intuition which we will develop as a community.

Comment (steveburnett) [reply CONTRIBUTING.md:125]: Works for me!

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md:493]: The indent is a suggestion on what you should do instead, I will add some language around that

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md:133]: Added

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md:125]: Integrated

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md]: added and adjusted some language

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md:153]: added elsewhere with link to RFC

File: CONTRIBUTING.md
Comment (FaultyJuggler) [reply CONTRIBUTING.md]: instructed them to follow the PR template provided

File: CONTRIBUTING.md
Comment (steveburnett) [inline CONTRIBUTING.md:492]:
```suggestion
    * Instead, review the related code, then draft initial documentation as a separate commit
```
Typo and light revision

File: CONTRIBUTING.md
Comment (tdcmeehan) [inline CONTRIBUTING.md:498]: What is lower tier support?

Comment (FaultyJuggler) [reply CONTRIBUTING.md:498]: I believe when we wrote this we were thinking about Tier 1,2,3
On a spectrum of community support, technical developer advocate, software engineer, senior software engineer
Each of these can delegate one level down
As an example, if someone submits a very complex commit that eventually will need a senior engineer to review, but the code style or lack of documentation etc. is hindering making it an efficient use of time, then it would get kicked back to DA or a community support person.

Publicly we may not need to mention this as this is more of a concern for senior engineers who belong to organizations that are part of the foundation, and not necessarily a public contributor who wants to volunteer time for code reviews.

Comment (tdcmeehan) [reply CONTRIBUTING.md:498]: Sounds good, agreed we can keep it out of the public doc

**Comment by tdcmeehan** [comment]
I see a few remaining comments from @steveburnett and @kiersten-stokes , otherwise LGTM once commits are squashed

**Comment by tdcmeehan** [comment]
Reopened in #21988

**Pull request closed without merging**
